core/src/main/java/org/infinispan/util/concurrent/jdk8backported/StrippedConcurrentLinkedDeque.java

```diff
@@ -0,0 +1,723 @@
+    * Non-invariants:
+    * - tail.item may or may not be null
+    * - tail may not be reachable from the first or last node, or from head
+    */
+   transient volatile DequeNode<E> tail;
+   
+   static final DequeNode<Object> PREV_TERMINATOR, NEXT_TERMINATOR;
+
+   final DequeNode<E> DEFAULT;
+
+   @SuppressWarnings("unchecked")
+   DequeNode<E> prevTerminator() {
+       return (DequeNode<E>) PREV_TERMINATOR;
+   }
+
+   @SuppressWarnings("unchecked")
+   DequeNode<E> nextTerminator() {
+       return (DequeNode<E>) NEXT_TERMINATOR;
+   }
+
+   DequeNode<E> defaultNode() {
+       return (DequeNode<E>) DEFAULT;
+   }
+
+   public StrippedConcurrentLinkedDeque() {
+      DEFAULT = new DequeNode<E>();
+      head = tail = defaultNode();
+   }
+
+   static class DequeNode<E> {
+       volatile DequeNode<E> prev;
+       volatile E item;
+       volatile DequeNode<E> next;
+
+       DequeNode() {  // default constructor for NEXT_TERMINATOR, PREV_TERMINATOR
+       }
+
+       @Override
+      public String toString() {
+         return "DequeNode [item=" + item + "]";
+      }
+
+      /**
+        * Constructs a new node.  Uses relaxed write because item can
+        * only be seen after publication via casNext or casPrev.
+        */
+       DequeNode(E item) {
+           UNSAFE.putObject(this, itemOffset, item);
+       }
+
+       void resetLazily(E item) {
+          UNSAFE.putObject(this, itemOffset, item);
+          lazySetNext(null);
+          lazySetPrev(null);
+       }
+
+       boolean casItem(E cmp, E val) {
+           return UNSAFE.compareAndSwapObject(this, itemOffset, cmp, val);
+       }
+
+       void lazySetNext(DequeNode<E> val) {
+           UNSAFE.putOrderedObject(this, nextOffset, val);
+       }
+
+       boolean casNext(DequeNode<E> cmp, DequeNode<E> val) {
+           return UNSAFE.compareAndSwapObject(this, nextOffset, cmp, val);
+       }
+
+       void lazySetPrev(DequeNode<E> val) {
+           UNSAFE.putOrderedObject(this, prevOffset, val);
+       }
+
+       boolean casPrev(DequeNode<E> cmp, DequeNode<E> val) {
+           return UNSAFE.compareAndSwapObject(this, prevOffset, cmp, val);
+       }
+
+       // Unsafe mechanics
+
+       private static final sun.misc.Unsafe UNSAFE;
+       private static final long prevOffset;
+       private static final long itemOffset;
+       private static final long nextOffset;
+
+       static {
+           try {
+               UNSAFE = getUnsafe();
+               Class<?> k = DequeNode.class;
+               prevOffset = UNSAFE.objectFieldOffset
+                   (k.getDeclaredField("prev"));
+               itemOffset = UNSAFE.objectFieldOffset
+                   (k.getDeclaredField("item"));
+               nextOffset = UNSAFE.objectFieldOffset
+                   (k.getDeclaredField("next"));
+           } catch (Exception e) {
+               throw new Error(e);
+           }
+       }
+   }
+
+   abstract class AbstractItr implements Iterator<DequeNode<E>> {
```

Comment:
This iterator was tweaked to return the Node itself instead of the value.  This is only used in testing though.